File: scripts/lib/dnc-check.js
/**
 * Shared Do-Not-Contact (DNC) enforcement module
 * ALL email-sending scripts MUST use this before sending.
 * 
 * Features:
 * - Case-insensitive matching
 * - Matches primary email AND alt_emails
 * - Logs all skips for audit trail
 * - Safe default: if DNC list fails to load, blocks ALL sends (fail-closed)
 */

const fs = require('fs');
const path = require('path');

const DNC_LIST_PATH = path.join(
  process.env.HOME || '/Users/openclaw',
  '.openclaw/workspace/data/massdwell/sales/do-not-contact-list.json'
);

let _cachedEmails = null;
let _cacheTime = 0;
const CACHE_TTL_MS = 60000; // Re-read file every 60s

/**
 * Load all DNC emails (primary + alt_emails), lowercased.
 * Returns Set<string> of blocked emails.
 * Throws on file read error (fail-closed).
 */
function loadDNCEmails() {
  const now = Date.now();
  if (_cachedEmails && (now - _cacheTime) < CACHE_TTL_MS) {
    return _cachedEmails;
  }

  if (!fs.existsSync(DNC_LIST_PATH)) {
    console.error(`🚨 DNC list not found at ${DNC_LIST_PATH} — BLOCKING ALL SENDS (fail-closed)`);
    return null; // null = fail-closed
  }

  const data = JSON.parse(fs.readFileSync(DNC_LIST_PATH, 'utf8'));
  const emails = new Set();

  for (const contact of (data.contacts || [])) {
    if (contact.email) emails.add(contact.email.toLowerCase().trim());
    if (Array.isArray(contact.alt_emails)) {
      for (const alt of contact.alt_emails) {
        emails.add(alt.toLowerCase().trim());
      }
    }
  }

  _cachedEmails = emails;
  _cacheTime = now;
  return emails;
}

/**
 * Check if an email is on the DNC list.
 * @param {string} email - Email to check
 * @param {string} [context] - Calling script name for logging
 * @returns {boolean} true if blocked (DO NOT SEND)
 */
function isDNC(email, context = 'unknown') {
  if (!email) return true; // No email = don't send

  const dncEmails = loadDNCEmails();

  // Fail-closed: if list couldn't load, block everything
  if (dncEmails === null) {
    console.error(`🚨 [DNC:${context}] BLOCKING ${email} — DNC list failed to load (fail-closed)`);
    return true;
  }

  const normalized = email.toLowerCase().trim();
  if (dncEmails.has(normalized)) {
    console.log(`⏭️ [DNC:${context}] SKIPPING ${email} — on do-not-contact list`);
    return true;
  }

  return false;
}

/**
 * Filter an array of prospects/objects, removing DNC entries.
 * @param {Array} prospects - Array of objects with 'email' field
 * @param {string} [context] - Calling script name for logging
 * @returns {Array} Filtered array with DNC entries removed
 */
function filterDNC(prospects, context = 'unknown') {
  return prospects.filter(p => !isDNC(p.email, context));
}

module.exports = { isDNC, filterDNC, loadDNCEmails };

File: scripts/lib/dnc_check.py
"""
Shared Do-Not-Contact (DNC) enforcement module for Python scripts.
ALL email-sending scripts MUST use this before sending.

Features:
- Case-insensitive matching
- Matches primary email AND alt_emails
- Logs all skips for audit trail
- Safe default: if DNC list fails to load, blocks ALL sends (fail-closed)
"""

import json
import time
from pathlib import Path

DNC_LIST_PATH = Path.home() / '.openclaw/workspace/data/massdwell/sales/do-not-contact-list.json'

_cached_emails = None
_cache_time = 0
_CACHE_TTL = 60  # seconds


def load_dnc_emails():
    """Load all DNC emails (primary + alt_emails), lowercased. Returns set or None (fail-closed)."""
    global _cached_emails, _cache_time
    now = time.time()
    if _cached_emails is not None and (now - _cache_time) < _CACHE_TTL:
        return _cached_emails

    if not DNC_LIST_PATH.exists():
        print(f"🚨 DNC list not found at {DNC_LIST_PATH} — BLOCKING ALL SENDS (fail-closed)")
        return None

    try:
        data = json.loads(DNC_LIST_PATH.read_text())
        emails = set()
        for contact in data.get('contacts', []):
            if contact.get('email'):
                emails.add(contact['email'].lower().strip())
            for alt in contact.get('alt_emails', []):
                emails.add(alt.lower().strip())
        _cached_emails = emails
        _cache_time = now
        return emails
    except Exception as e:
        print(f"🚨 DNC list failed to load: {e} — BLOCKING ALL SENDS (fail-closed)")
        return None


def is_dnc(email, context='unknown'):
    """Check if email is on DNC list. Returns True if blocked (DO NOT SEND)."""
    if not email:
        return True

    dnc_emails = load_dnc_emails()
    if dnc_emails is None:
        print(f"🚨 [DNC:{context}] BLOCKING {email} — DNC list failed to load (fail-closed)")
        return True

    normalized = email.lower().strip()
    if normalized in dnc_emails:
        print(f"⏭️ [DNC:{context}] SKIPPING {email} — on do-not-contact list")
        return True
    return False


def filter_dnc(prospects, email_key='email', context='unknown'):
    """Filter a list of prospect dicts, removing DNC entries."""
    return [p for p in prospects if not is_dnc(p.get(email_key, ''), context)]
